programs/perpetuals/src/instructions/claim_stakes.rs

```diff
@@ -0,0 +1,411 @@
+        mut,
+        token::mint = lm_token_mint,
+        seeds = [b"staking_lm_reward_token_account"],
+        bump = cortex.staking_lm_reward_token_account_bump
+    )]
+    pub staking_lm_reward_token_account: Box<Account<'info, TokenAccount>>,
+
+    /// CHECK: empty PDA, authority for token accounts
+    #[account(
+        seeds = [b"transfer_authority"],
+        bump = perpetuals.transfer_authority_bump
+    )]
+    pub transfer_authority: AccountInfo<'info>,
+
+    #[account(
+        mut,
+        seeds = [b"staking",
+                 owner.key().as_ref()],
+        bump = staking.bump
+    )]
+    pub staking: Box<Account<'info, Staking>>,
+
+    #[account(
+        mut,
+        seeds = [b"cortex"],
+        bump = cortex.bump,
+        has_one = staking_reward_token_mint
+    )]
+    pub cortex: Box<Account<'info, Cortex>>,
+
+    #[account(
+        seeds = [b"perpetuals"],
+        bump = perpetuals.perpetuals_bump
+    )]
+    pub perpetuals: Box<Account<'info, Perpetuals>>,
+
+    #[account(
+        seeds = [b"lm_token_mint"],
+        bump = cortex.lm_token_bump
+    )]
+    pub lm_token_mint: Box<Account<'info, Mint>>,
+
+    #[account()]
+    pub staking_reward_token_mint: Box<Account<'info, Mint>>,
+
+    perpetuals_program: Program<'info, program::Perpetuals>,
+    system_program: Program<'info, System>,
+    token_program: Program<'info, Token>,
+}
+
+pub fn claim_stakes(ctx: Context<ClaimStakes>) -> Result<()> {
+    let staking = ctx.accounts.staking.as_mut();
+    let cortex = ctx.accounts.cortex.as_mut();
+
+    msg!("Process resolved rounds & rewards calculation");
+
+    // Loop over resolved rounds and:
+    // 1. Calculate rewards token amount to be claimed for staker
+    // 2. Drop fully claimed rounds
+    let (
+        rewards_token_amount,
+        stake_amount_with_reward_multiplier,
+        lm_rewards_token_amount,
+        stake_amount_with_lm_reward_multiplier,
+    ) = {
+        // prints compute budget before
+        sol_log_compute_units();
+
+        let resolved_staking_rounds_len_before = cortex.resolved_staking_rounds.len();
+        let stake_token_decimals = cortex.stake_token_decimals as i32;
+        let stake_reward_token_decimals = cortex.stake_reward_token_decimals as i32;
+
+        let mut rewards_token_amount: u64 = 0;
+        let mut lm_rewards_token_amount: u64 = 0;
+
+        // total amount of token that tokens have been claimed for
+        let mut stake_amount_with_reward_multiplier: u64 = 0;
+        let mut stake_amount_with_lm_reward_multiplier: u64 = 0;
+
+        msg!(
+            "{} resolved rounds to evaluate",
+            cortex.resolved_staking_rounds.len()
+        );
+
+        // For each resolved staking rounds
+        cortex.resolved_staking_rounds.retain_mut(|round| {
+            // Calculate rewards for locked stakes
+            {
+                // For each user locked stakes
+                for locked_stake in staking.locked_stakes.iter_mut() {
```

Comment:
Wondering where is the computing limit for that loop in term of amount of resolved staking round/stakes. We might want to find this our with tests